Parsedown.php

```diff
@@ -1488,7 +1488,21 @@ protected function element(array $Element)
             }
         }
 
+        $unsafeHtml = false;
         if (isset($Element['text']))
+        {
+            $text = $Element['text'];
+        }
+        // very strongly consider an alternative if you're writing an
+        // extension
+        elseif (isset($Element['unsafeHtml']))
+        {
+            $text = $Element['unsafeHtml'];
```

Comment:
Calling this `unsafeHtml` will result in the exact opposite of what your intentions were :wink: 

If one reads `$Element['name']`, one will put the element's name in there. If one reads `$Element['unsafeHtml']`, one will put unsafe HTML in there. :laughing: This should rather be called `safeHtml`, `trustedHtml` or, preferrably, just `html`. Parsedown is a library, we can assume that developers read the manual.

Reply:
I was going for more along the lines of https://developer.apple.com/documentation/swift/unsafepointer. I think it is clear that `unsafeHtml` contains unsafe HTML, whether or not putting stuff in there is a good idea or not depends how much you trust it ;p

(I think the best name is probably going to depend on whether there is an assumption that we will automatically escape something called `unsafeHtml`).

I think that whenever this feature is used, the HTML is unsafe (i.e. it depends on the trustworthiness of input). If you know enough about the input at runtime for it to be safe, you will be able to refactor to the preferred AST representation.

Reply:
> I was going for more along the lines of https://developer.apple.com/documentation/swift/unsafepointer. I think it is clear that `unsafeHtml` contains unsafe HTML

Having a `UnsafePointer` is something completely different, it's a "unsafe *pointer* to something". A array key in contrary is no pointer, it's rather a variable *containing* a value. That's the fundamental difference between pointers and variables in programming. A variable's name, and a array key is nothing else than that, must describe the variable's value. Don't confuse variables with pointers!

Furthermore, `unsafeHtml` contains no unsafe HTML - this statement is simply wrong. First of all it contains just HTML - whether it is safe or unsafe depends on what the extension's developer puts in there. However, we explicitly tell extension developers that they are supposed to put *safe HTML* in there, i.e. if the developer puts unsafe HTML in there, he is doing it wrong. Whichever way you look at it, it should never be "unsafe HTML".

> If you know enough about the input at runtime for it to be safe, you will be able to refactor to the preferred AST representation.

#568 is a perfect example that this assumption isn't true :wink:

Reply:
Something that is important to me is that whatever name we pick hints at the following:
> [feature] provides no automated escaping or sanitisation. You are responsible for handling the escaping and sanitisation of any HTML you work with through [feature] to avoid cross-site scripting.

If "unsafeHTML" doesn't do that, fine. Let's figure out something that does.

It's more important that you understand *why* I have chosen this name, it is not important to me that you agree with it because the name isn't fixed. If the name is doing a bad job at explaining the why then let's change it.

---

> Having a `UnsafePointer` is something completely different, it's a "unsafe pointer to something". A array key in contrary is no pointer, it's rather a variable containing a value. That's the fundamental difference between pointers and variables in programming. A variable's name, and a array key is nothing else than that, must describe the variable's value. Don't confuse variables with pointers!

Thanks for the explainer 😒, but I'm not trying to compare an array key to a pointer. This was purely a linguistic comparison of the use of the word "unsafe".

This was a response to:

> Calling this unsafeHtml will result in the exact opposite of what your intentions were

The link was a citation to justify the choice of name. In that, in the link unsafe[Thing] is used to mean that "you should be careful when using [Thing]", because there are known examples of unsafe behaviour. The same is true in our example. If you are not careful with the HTML being written into the variable (for example if you write untrusted user-input into it), you will yield unsafe results.

Again, the only purpose of the link was to cite an example of the word "unsafe" being used to mean "be careful", and not "don't be careful". 

---

> First of all it contains just HTML - whether it is safe or unsafe depends on what the extension's developer puts in there. However, we explicitly tell extension developers that they are supposed to put safe HTML in there

I agree with all of this, my intention is not to instruct developers to put unsafe HTML in this location, but rather to use the word as an explicit reminder that what they are doing can have unsafe results and they must take care (i.e. a "be careful" and not "don't be careful").

Here's what the documentation I linked to states about unsafe[Thing]:
> UnsafePointer provides no automated memory management or alignment guarantees. You are responsible for handling the life cycle of any memory you work with through unsafe pointers to avoid leaks or undefined behavior.

Here's basically the same description, replaced with things that can go wrong in HTML land, rather than in memory:
> Unsafe ***HTML*** provides no automated ***escaping*** or ***sanitisation***. You are responsible for handling the ***escaping and sanitisation*** of any ***HTML*** you work with through unsafe ***HTML*** to avoid ***cross-site scripting***.

I suppose this really uses "unsafe" to mean "there is no safety provided for you, please provide it yourself", rather than "this is where you should put things that are unsafe".

You actually get the reverse problem if we call give it a positive prefix, e.g. something like "safeHtml", does this indicate that we are applying sanitisation to make it safe? The arguments for misinterpretation go both ways round.